File: 0383-ransom-note/0383-ransom-note.py
class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        char_in_mag = {}
        char_in_ransomNote = {}

        for char in magazine:
            if char in char_in_mag:
                char_in_mag[char] += 1
            else:
                char_in_mag[char] = 1

        for char in ransomNote:
            if char in char_in_ransomNote:
                char_in_ransomNote[char] += 1
            else:
                char_in_ransomNote[char] = 1

        for char in ransomNote:
            if char not in char_in_mag or char_in_mag[char] < char_in_ransomNote[char]:
                return False

        return True

File: 0383-ransom-note/0383-ransom-note.java
class Solution {
    public boolean canConstruct(String ransomNote, String magazine) {

        HashMap<Character, Integer> char_in_mag = new HashMap<>();
        HashMap<Character, Integer> char_in_ransomNote = new HashMap<>();

        for (char c : magazine.toCharArray()) {
            if (char_in_mag.containsKey(c)) {
                char_in_mag.put(c, char_in_mag.get(c) + 1);
            } else {
                char_in_mag.put(c, 1);
            }
        }

        for (char c : ransomNote.toCharArray()) {
            if (char_in_ransomNote.containsKey(c)) {
                char_in_ransomNote.put(c, char_in_ransomNote.get(c) + 1);
            } else {
                char_in_ransomNote.put(c, 1);
            }
        }

        for (char c : ransomNote.toCharArray()) {
            if (!char_in_mag.containsKey(c) || char_in_mag.get(c) < char_in_ransomNote.get(c)) {
                return false;
            }
        }

        return true;
    }
}
